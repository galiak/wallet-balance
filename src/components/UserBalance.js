import React from 'react';
import {connect} from 'react-redux';
import {deposit, withdraw} from '../actions/balance';

export class UserBalance extends React.Component {
  state = {balance: undefined};

  update = event => this.setState({balance: parseInt(event.target.value, 10)});

  deposit = () => this.props.deposit(this.state.balance);

  withdraw = () => this.props.withdraw(this.state.balance);

  render() {
    return (
      <div> 
        <h3 className="balance">Balance: {this.props.balance}</h3>
        <div className="ui input">
          <input className="input-wallet ui input" onChange={this.update} />
          <div className="ui buttons">
            <button className="btn-deposit ui positive button" onClick={this.deposit}>Deposit</button>
            <div className="or"></div>
            <button className="btn-withdraw ui button" onClick={this.withdraw}>Withdraw</button>
          </div>
        </div>
      </div>
    );
  }
}

export default connect(state => {return {balance: state.balance}}, {deposit, withdraw})(UserBalance);