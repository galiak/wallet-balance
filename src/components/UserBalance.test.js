import React from 'react';
import {shallow} from 'enzyme';
import {UserBalance} from './UserBalance';
import '../setupTests';

describe('UserBalance', () => {
  const mockDeposit = jest.fn();
  const mockWithdraw = jest.fn();
  const props = {balance: 20, deposit: mockDeposit, withdraw: mockWithdraw};
  const wallet = shallow(<UserBalance {...props} />);
  
  it('renders properly', () => {
    expect(wallet).toMatchSnapshot();
  });

  it('displays the balance from props', () => {
    expect(wallet.find('.balance').text()).toEqual('Balance: 20');
  });

  it('create an input to deposit/withdraw from the balance', () => {
    expect(wallet.find('.input-wallet').exists()).toBe(true);
  });

  describe('when a user types in the wallet input', () => {
    const userBalance = '25';
    
    beforeEach(() => {
      wallet
        .find('.input-wallet')
        .simulate('change', {target: {value: userBalance}});
    });

    it('updates the local balance in `state` and converts it to a number', () => {
      expect(wallet.state().balance).toEqual(parseInt(userBalance, 10));
    });

    describe('and the user wants to make a deposit', () => {
      beforeEach(() => {
        wallet.find('.btn-deposit').simulate('click');
      });

      it('dispatches the `deposit()` it receives from props with the local balance', () => {
        expect(mockDeposit).toHaveBeenCalledWith(parseInt(userBalance, 10));
      });
    });

    describe('and the user wants to withdraw', () => {
      beforeEach(() => {
        wallet.find('.btn-withdraw').simulate('click');
      });

      it('dispatches the `withdraw()` it receives from props with the local balance', () => {
        expect(mockWithdraw).toHaveBeenCalledWith(parseInt(userBalance, 10));
      });
    });
  });
});