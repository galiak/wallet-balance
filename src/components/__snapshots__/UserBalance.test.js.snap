// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserBalance renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <UserBalance
    balance={20}
    deposit={[MockFunction]}
    withdraw={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3
          className="balance"
        >
          Balance: 
          20
        </h3>,
        <input
          className="input-wallet ui input"
          onChange={[Function]}
        />,
        <button
          className="btn-deposit ui button"
          onClick={[Function]}
        >
          Deposit
        </button>,
        <button
          className="btn-withdraw ui button"
          onClick={[Function]}
        >
          Withdraw
        </button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Balance: ",
            20,
          ],
          "className": "balance",
        },
        "ref": null,
        "rendered": Array [
          "Balance: ",
          20,
        ],
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "input-wallet ui input",
          "onChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Deposit",
          "className": "btn-deposit ui button",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Deposit",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Withdraw",
          "className": "btn-withdraw ui button",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Withdraw",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3
            className="balance"
          >
            Balance: 
            20
          </h3>,
          <input
            className="input-wallet ui input"
            onChange={[Function]}
          />,
          <button
            className="btn-deposit ui button"
            onClick={[Function]}
          >
            Deposit
          </button>,
          <button
            className="btn-withdraw ui button"
            onClick={[Function]}
          >
            Withdraw
          </button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Balance: ",
              20,
            ],
            "className": "balance",
          },
          "ref": null,
          "rendered": Array [
            "Balance: ",
            20,
          ],
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "input-wallet ui input",
            "onChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Deposit",
            "className": "btn-deposit ui button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Deposit",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Withdraw",
            "className": "btn-withdraw ui button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Withdraw",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
